### ISSUE TITLE
there's no way to get a shell from a build or vice versa

### ISSUE THREAD

**Issue by dnr** [tweag/npmlock2nix#51]
**Is your feature request related to a problem? Please describe.**

I have a project with a node-based build, so I have a `npmlock2nix.build` expression in my defaut.nix. Of course, I'd also like to run parts of it interactively and incrementally, so I have a `npmlock2nix.shell` expression in my shell.nix. This seems redundant: I have to duplicate the `src` attribute, at least.

The structure of the code suggests that since `.build` and `.shell` both depend on `.node_modules`, which has the meat of the logic, I should call `.node_modules` and then create the `build` and `shell` from that. But there's no way to actually do that.

**Describe the solution you'd like**

It seems like the structure is inverted: instead of exposing three functions `node_modules`, `shell`, and `build`, which take similar attributes, there should be a single function that takes `src` and various other attributes describing the build and dependencies, and then outputs a set with three attributes `build`, `shell`, and `node_modules`.

Alternately, functions that create a shell from a node_modules, or a build from a node_modules, would address this. Though I think that a single function that exposes all of them at once is probably a cleaner solution.

**Comment by dnr** [comment]
I tried this out here: https://github.com/dnr/npmlock2nix/tree/reorganize

It needs some more polishing but it works for me, and personally I find this API easier to understand: The attributes for the three derivations (node_modules, shell, and build) are more cleanly separated, so there's no need to remember which ones affect which derivation, and no need for filtering attributes in the implementation. It's easy to get both a shell and a build derivation from the same "configuration" (the original goal).

I've only tried this with one project so there might be some subtleties that I'm missing (esp. around native dependencies). And there are still some open questions, like should `buildInputs`/`nativeBuildInputs` apply to all three derivations or just some? Anyway, let me know if this makes sense and if you want me to clean it up for a PR.

**Comment by andir** [comment]
You are right that right now the API docs aren't that great in that regard. My initial idea was that it would actually compose as you described. For some reason (that has escaped my mind) we are now *always* using `src` in the `node_modules` derivation while the initial idea was to just use `src` to populate the path to the `package.json` and `package-lock.json`: https://github.com/tweag/npmlock2nix/blob/1499b402eda50976435f5b1b0525dc7d0404900b/internal.nix#L226-L227

We could perhaps provide a `setup` function that exposes all the various kinds for a given project. I'd really like to keep the small composable function as it gives you (the users) plenty of ways to customize a build if required. For that to actually work out in practice we will have to allow passing in `node_modules` into the `build` & `shell` function again.

The attributes buildInputs & nativeBuildInputs should only be specific to the specific build phase. While producing `node_modules` might require a GCC on path the actual build of (most) Node projects probably doesn't and shouldn't have it in the environment.

Regardless of the above I pretty much like how you finally fixed the oddness arround `lower_cased` attribute names and changed them to `camelCase` instead. We should pick that up as well. Seems more consistent with nixpkgs.

**Comment by dnr** [comment]
It's ironic because I prefer `lower_with_underscores` style to `camelCase`, but I prefer consistency even more :)

I'm not actively working on my node project at the moment, so I'm not likely to do any more work on this myself, but I'll look out for developments here and try them out.
